### ISSUE TITLE
Parse inline markup during the parse phase (use recursive descent parser)

### ISSUE THREAD

**Issue by mojavelinux** [asciidoctor/asciidoctor#61]
Currently, the parsing of inline content happens during rendering. This limits the information you have in the document after parsing. Instead, Asciidoctor should parse all text extents into inline nodes during parsing.

This requires moving the substitutions from the rendering phase to the parsing phase.

It also means that each line in the buffer will become an array of inline nodes that represent the chunked text, toggling between plain text and elements like links and images.

Since this has the potential to slow down a single pass parse/render, we may want to have a flag which controls the phase in which inline text is parsed.

**Comment by benignbala** [comment]
Should I have a go at this using ANTLR4 (https://rubygems.org/gems/antlr4/versions/0.9.2) ? Thanks

**Comment by benignbala** [comment]
Or rather, shall we try it with ANTLR4 in Java in the asciidoctorj project ?

**Comment by mojavelinux** [comment]
There's a project setup w/ ANTLR to do this experimentation. See https://github.com/asciidoctor/asciidoc-grammar-prototype

**Comment by mojavelinux** [comment]
Test case:

The following should not be matched as emphasis (i.e., italics):

```
*_id* word_
```

**Comment by elextr** [comment]
I provided advice to another project that was trying to develop an Asciidoc implementation in C++ for an embedded project.  That project has now been dropped (the embedded project and therefore the Asciidoc implementation :disappointed: ), but here are some lessons for future implementers of this parser.

Using the traditional parser tools is problematical, they all assume context free grammers, and Asciidoc is definitely not a CFG.  Even the tools various ways of handling contextual dependence did not seem sufficient for this case (YMMV but don't waste too much time on it, so far nobody made it work).  The example in the post above is a simple context dependence (parsing the `_id` as italic should stop if it hits the end of any quotes it is nested in, thats the context it depends on).

The big context dependence is of course the the different parsing between blocks (and changing that with the `subs=` option, since it changes the context dependence dynamically during parsing).  So you have to recognise the block type or `subs` option early and apply it to the appropriate part of the input, even before you have parsed the appropriate input.  

An alternative that was explored was to parse everything and then just revert the parses that were not required back to text.  Unfortunately what is parsed affects how other things parse, eg if quotes are not parsed because the opening is inside a macro target and the ending is outside that will change if subsequently its discovered that macros are not in the `subs=` list, eg
```
[subs=-macros]
the italicised part of http:xxx__yyy[]__ frobnicates the foo bar
```

Its not entirely bad news, the top level structure of a document (sections, blocks, lists etc) is a CFG and defines most of the context for parsing the lower levels, so it seems possible to use a two pass approach where the structure and attribute lists are parsed and then then the block contents are parsed based on the context that specifies.

Thats where it was at when things stopped.  Hope this will help future implementers avoid many of the same issues (I'm sure there are plenty still left for them to find).

**Comment by mojavelinux** [comment]
Thanks for this input. I'll study it in detail.

Just to share the idea in my head, I am definitely not planning to implement it as a single parser / grammar. My idea was to do the selective parsing I am doing now, which is working fantastic, but switch to the parser / grammar for inline stuff where the context is fixed. The problem right now is only there. The rest of AsciiDoc is actually easy to parse using a line-based approach.

I haven't validated that idea yet, but from my experience, I'm very confident it's going to work.

**Comment by elextr** [comment]
Well, the inline stuff is in fact the part that depends on the context its in (see the `subs=` problem).

A hand carved RD or other parser that passes lots of context down to parsing of nested constructs is probably possible, if messy due to all the context tests.  But due to the context dependence it can't be memoised I don't think, so it will likely have the dreaded exponential time performance.  Though that may not be a problem if it is constrained to previously determined limited segments of the input as you propose.

**Comment by benignbala** [comment]
@mojavelinux, But there are several benefits to having a complete (multi-pass) state based parser in addition to the streaming parser that asciidoctor currently has.  The main benefit being, once we have the complete AST, it becomes very simple to translate from AsciiDoctor to any other output format (esp. thinking of ReST to make adoc files generate .rst files for sphinx). So, as @elextr suggests, a hand-written recursive descent is a nicer soltuion, although performance wise, it will be a lot slower. 

My idea is to have the state transitions through generated code and a JSON/XML file driving the syntax/grammar logic, so that we can add more the parser API in more languages by adding necessary templates for the specific language.

**Comment by andya9** [comment]
@benignbala IMO conversion to other formats can be easily done like this: asciidoc > DOCBOOK > anything else using the huge existing ecosystem for dealing with docbook and xml in general. In other words, docbook could already be considered a viable AST

**Comment by jirutka** [comment]
My idea is to write a standalone AsciiDoc parser in PEG that would output an AST represented as JSON. This would strictly decouple parser and renderer.

**Comment by Mogztter** [comment]
a Rust parser for AsciiDoc would be awesome! :+1:

**Comment by Mogztter** [comment]
@jirutka Did you consider https://github.com/Geal/nom ?

**Comment by jirutka** [comment]
Yes, I know about nom, but haven't tried it yet.

**Comment by s-leroux** [comment]
Hi @mojavelinux,
What is the current status of this issue? Is it still WIP or has the idea being dropped?

Two reasons I ask that for:
1. The docs make explicit reference to that issue in the ["Extension" section](https://asciidoctor.org/docs/user-manual/#extensions)
2. :) I would find that very useful to extract link targets from the AST to check for broken links before rendering AsciiDoctor documents.

**Comment by mojavelinux** [comment]
Definitely not dropped. The best way to describe the current state is "not enough money to work on it yet".

I completely understand your use case. You may also be interested in the fact that I'm starting to work on a validator for AsciiDoc over here, which by its very nature has to parse inline elements. (see https://github.com/opendevise/textlint-plugin-asciidoc). This will be essential for Antora.

Before we can put an inline parser in Asciidoctor, we have to write one in a lab to find out the unknowables. Will there be decisions we have to make about the syntax in order to make the switch? Will there be a migration? etc, etc. I'm coming up very rapidly on needing an inline parser for a variety of use cases, so it is still very high on my list of efforts to start.

**Comment by mojavelinux** [comment]
@elextr That's great to hear!

What I know for sure is that there's no way a parser could process the whole document at once. AsciiDoc is too contextual for that. What needs to happen is that the blocks have to be identified first (as it descends), which is the approach I've started to use in the validator parser. (This is easily done by hand). Then, based on the block context and subs overrides, the appropriate inline parser is selected to descend into the block to process the inline nodes. (This is where a PEG parser comes in). In the end, it's a hybrid solution. I believe that will give us exactly the results we want.

**Comment by mojavelinux** [comment]
I also want to point out that we're not shooting for something pure. All we're shooting for is something that doesn't use multiple regexp/gsub passes over the same text (and which gives us an AST).

**Comment by elextr** [comment]
@mojavelinux yes, parsing structural markup first certainly helps, its one of the things I tried in the past. 
 But inline parsing still has problems.

An example that caused me headaches.  (Note I love PEG parsers, I have made several over the years, but it would need some sort of extended capability beyond normal PEG to handle the following).

```
:blah__blah: foo

[subs=attributes]
blah{blah__blah}blah__
```

vs

```
:blah__blah: foo

[subs=quotes]
blah{blah__blah}blah__
```

Same inline markup, but different AST depending on the subs value.  So your PEG has to depend on that somehow.

And there is no simple AST that can record both the attribute and the quotes at the same time since they overlap.  Fuzzy parsers and ASTs with multiple options that are later collapsed are of course an option, but they are not simple PEG parsers.

And then the question of what that resolves to in the absence of an explicit `subs=` attribute to limit it to either `quotes` or `attributes`, by default `subs` has everything set.  Asciidoctor resolves it in favour of quotes due to a side effect of order of substitution.

But the first token is the attribute token `{` so a "real" parser will recognise the attribute and not recognise the quotes (and must not choke on the unmatched quotes as well) which is the opposite behaviour from current implementations.

To date I am just taking the "real parsers are different" route and will see how much difference it makes in the real world (not much I suspect).

**Comment by mojavelinux** [comment]
> Same inline markup, but different AST depending on the subs value. So your PEG has to depend on that somehow.

Easy. You just use a different PEG for the two cases. Not necessarily a traditional thing to do, but what parsing AsciiDoc calls for. That's what I mean when I say a hybrid parser.

**Comment by mojavelinux** [comment]
> But the first token is the attribute token { so a "real" parser will recognise the attribute and not recognise the quotes

I think either attributes have to be replaced first, or we have to change the rules about boundaries. I agree this is something that will require more thought.

**Comment by mojavelinux** [comment]
On second look, I understand better what you're pointing out. Yes, I agree there are going to be differences in the edge cases. These changes in the syntax are going to have to happen if we expect to move past this problem. We have to evolve.

**Comment by elextr** [comment]
> Easy. You just use a different PEG for the two cases. 

Lets see, six binary options for `subs=` thats 64 combinations, so 64 parsers, I suppose that is manageable.  There might be some sharing of common parts as well.

> Not necessarily a traditional thing to do, but what parsing AsciiDoc calls for.

Indeed, Asciidoc is not a context free grammar, so traditional context free techniques should not be expected to work.

Its good to see that you understand that a simple PEG parser isn't the solution by itself, I guess I harp on it a bit because many previous posts on the topic by various people seem to suggest its all you need, and it would be sad to see people wasting their time pursuing that path.

**Comment by s-leroux** [comment]
@mojavelinux, @elextr Great to hear things are things are evolving concerning this issue.

> These changes in the syntax are going to have to happen if we expect to move past this problem. We have to evolve.

If I understand it well, we may have to expect syntax change in AsciiDoctor source document _before_ we can see inline nodes parsed into the AST? Do you have already some ideas in mind? Or maybe is there already another issue to discuss those changes?

**Comment by mojavelinux** [comment]
> If I understand it well, we may have to expect syntax change in AsciiDoctor source document before we can see inline nodes parsed into the AST?

The goal is to minimize the impact as much as possible, but there may be slight differences at the edge cases. We won't really know until we get it fully working.

**Comment by mojavelinux** [comment]
For those following along, @Mogztter has been working on an inline parser to put theory into practice, tease out unknowns, and move the conversation forward. I still owe him feedback, but perhaps others are interested in weighing in.

See https://github.com/Mogztter/asciidoctor-inline-parser

**Comment by vmassol** [comment]
Thanks for the heads up @mojavelinux 

FTR, on the XWiki side we use java so we would need some java version to try it out/provide feedback/help.

Have a great weekend.

**Comment by laubai** [comment]
Found my way here from https://github.com/asciidoctor/asciidoctor/issues/1678. We ran into this, and https://github.com/asciidoctor/asciidoctor/issues/2600, but while that one is fixed in the current version, this one is not yet, so wanted to provide some relevant test cases.

Hope this helps, apologies for noise if it doesn't.

[master_adoc.txt](https://github.com/asciidoctor/asciidoctor/files/2743659/master_adoc.txt)

**Comment by bripmccann** [comment]
Typo in the title here ('descent' misspelled as 'decent')

**Comment by janderssonse** [comment]
I just ended up here, started with looking at Asciidoc-support in docusaurs, ended up looking at inline parser for ast much like is done for markdown in the unified/mdast/remark eco-system and learning the details I understand it is not so simple. Anyway, This fresh implementation  in haskell as an pandoc front https://github.com/gmarpons/asciidoc-hs looks interesting, but I guess it is already well known in the asciidoc community.

**Comment by mojavelinux** [comment]
Yes, we're aware of that implementation and Guillem's work. We are hopeful it will be instrumental in helping us to make the transition to an inline parser in the AsciiDoc Language specification. See https://gitlab.eclipse.org/eclipse/asciidoc/asciidoc-lang/-/issues/8 for the first step towards that goal. Asciidoctor will not implement an inline parser until it is specified in the AsciiDoc language project because it could have important implications for the parsing rules and thus must be resolved there first.
